homeassistant/components/sensor/history_stats.py

```diff
@@ -0,0 +1,314 @@
+
+        self._period = (datetime.datetime.now(), datetime.datetime.now())
+        self.value = 0
+
+        # noinspection PyUnusedLocal
+        # pylint: disable=invalid-name
+        @callback
+        def async_stats_sensor_state_listener(entity, old_state, new_state):
+            """Called when the sensor changes state."""
+            hass.async_add_job(self.async_update_ha_state, True)
+
+        async_track_state_change(
+            hass, entity_id, async_stats_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return round(self.value, 2)
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit the value is expressed in."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """Polling required."""
+        return True
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        start, end = self._period
+        hsh = HistoryStatsHelper
+        return {
+            ATTR_VALUE: hsh.pretty_duration(self.value),
+            ATTR_RATIO: hsh.pretty_ratio(self.value, self._period),
+            ATTR_START: start.strftime('%Y-%m-%d %H:%M:%S'),
+            ATTR_END: end.strftime('%Y-%m-%d %H:%M:%S'),
+        }
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return ICON
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Get the latest data and updates the states."""
+        # Parse templates
+        self.update_period()
+        start, end = self._period
+        # Convert to UTC
+        start = dt_util.as_utc(start)
+        end = dt_util.as_utc(end)
+
+        if not HistoryStatsHelper.wait_till_db_ready():
+            _LOGGER.error('Cannot connect to database')
+            return
+
+        # Get history between start and end
+        history_list = history.state_changes_during_period(
+            start, end, str(self._entity_id))
+
+        if self._entity_id not in history_list.keys():
+            return
+
+        # Get the first state
+        last_state = history.get_state(start, self._entity_id)
+        last_state = (last_state is not None and
+                      last_state == self._entity_state)
+        last_time = dt_util.as_timestamp(start)
+        elapsed = 0
+
+        # Make calculations
+        for item in history_list.get(self._entity_id):
+            current_state = item.state == self._entity_state
+            current_time = item.last_changed.timestamp()
+
+            if last_state:
+                elapsed += current_time - last_time
+
+            last_state = current_state
+            last_time = current_time
+
+        # Save value in hours
+        self.value = elapsed / 3600
+
+    def update_period(self):
```

Comment:
If this is an async method, mark it as such with either `@callback` or `@asyncio.coroutine`. Also prefix the name with `async_`